builtInFunctions/setGuardian.go

```diff
@@ -0,0 +1,270 @@
+// Case 5. User has TWO guardians. FIRST is enabled, SECOND is enabled => replace oldest one + set new one as pending
+func (sg *setGuardian) ProcessBuiltinFunction(
+	senderAccount, receiverAccount vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+) (*vmcommon.VMOutput, error) {
+	sg.mutExecution.RLock()
+	defer sg.mutExecution.RUnlock()
+
+	err := sg.checkArguments(senderAccount, receiverAccount, vmInput)
+	if err != nil {
+		return nil, err
+	}
+
+	guardians, err := sg.guardians(senderAccount)
+	if err != nil {
+		return nil, err
+	}
+	if sg.contains(guardians, vmInput.Arguments[0]) {
+		return nil, ErrGuardianAlreadyExists
+	}
+
+	switch len(guardians.Data) {
+	case 0:
+		// Case 1
+		return sg.tryAddGuardian(senderAccount, vmInput.Arguments[0], guardians, vmInput.GasProvided)
+	case 1:
+		// Case 2
+		if sg.pending(guardians.Data[0]) {
+			return nil, fmt.Errorf("%w: %s", ErrOwnerAlreadyHasOneGuardianPending, sg.pubKeyConverter.Encode(guardians.Data[0].Address))
+		}
+		// Case 3
+		return sg.tryAddGuardian(senderAccount, vmInput.Arguments[0], guardians, vmInput.GasProvided)
+	case 2:
+		// Case 4
+		if sg.pending(guardians.Data[1]) {
+			return nil, fmt.Errorf("%w: %s", ErrOwnerAlreadyHasOneGuardianPending, sg.pubKeyConverter.Encode(guardians.Data[1].Address))
+		}
+		// Case 5
+		guardians.Data = guardians.Data[1:] // remove oldest guardian
+		return sg.tryAddGuardian(senderAccount, vmInput.Arguments[0], guardians, vmInput.GasProvided)
+	default:
+		return &vmcommon.VMOutput{ReturnCode: vmcommon.ExecutionFailed}, nil
+	}
+}
+
+func (sg *setGuardian) checkArguments(
+	senderAccount,
+	receiverAccount vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+) error {
+	if check.IfNil(senderAccount) {
+		return fmt.Errorf("%w for sender", ErrNilUserAccount)
+	}
+	if check.IfNil(receiverAccount) {
+		return fmt.Errorf("%w for receiver", ErrNilUserAccount)
+	}
+	if vmInput == nil {
+		return ErrNilVmInput
+	}
+	if !(bytes.Equal(senderAccount.AddressBytes(), receiverAccount.AddressBytes()) &&
+		bytes.Equal(senderAccount.AddressBytes(), vmInput.CallerAddr)) {
+		return ErrOperationNotPermitted
+	}
+	if vmInput.CallValue == nil {
+		return ErrNilValue
+	}
+	if !isZero(vmInput.CallValue) {
+		return ErrBuiltInFunctionCalledWithValue
+	}
+	if len(vmInput.Arguments) != 1 {
+		return fmt.Errorf("%w, expected 1, got %d ", ErrInvalidNumberOfArguments, len(vmInput.Arguments))
+	}
+	if !sg.isAddressValid(vmInput.Arguments[0]) {
+		return fmt.Errorf("%w for guardian", ErrInvalidAddress)
+	}
+	if bytes.Equal(vmInput.CallerAddr, vmInput.Arguments[0]) {
+		return ErrCannotOwnAddressAsGuardian
+	}
+	if vmInput.GasProvided < sg.funcGasCost {
+		return ErrNotEnoughGas
+	}
+
+	return nil
+}
+
+func isZero(n *big.Int) bool {
+	return len(n.Bits()) == 0
+}
+
+func (sg *setGuardian) isAddressValid(addressBytes []byte) bool {
+	isLengthOk := len(addressBytes) == sg.pubKeyConverter.Len()
```

Comment:
you do not need address converter - address does not need to be given in bech32 format - it can be given directly as hex.encoded(bytes).

Reply:
length check can be done via len(senderAccount.Address()) == len(addressBytes)

Reply:
Right, I have:
1. Removed `pubKeyConverter` and replaced `pubKeyConverter.Decode` ->  `hex.EncodeToString`
2. Added len check via `len(senderAccount.Address()) == len(addressBytes)`